### PR TITLE
itm: init at 0.3.1

### PR DESCRIPTION
**Pull request by hazel-ocean**
###### Motivation for this change

I was working through the [rust-embedded/discovery](https://github.com/rust-embedded/discovery) book which calls for this specific tool. I was packaging it locally but figured I would upstream the package to NixOS as there are probably others who would find value in this tool being available.

One thing I noticed is that the `Cargo.lock` file was out of date for the 0.3.1 release which Nix was quite angered by, so I ran `cargo generate-lockfile` to circumvent the issue.

I also had to remove some `#![deny(warnings)]` from the codebase as Rust emits warnings for deprecated language features which was resulting in compilation errors.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by legendofmiracles** [comment]
Result of `nixpkgs-review pr 143750` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>itm</li>
  </ul>
</details>

error: hash mismatch in fixed-output derivation '/nix/store/cklr7y4bagvyhnk8d2f78rhz1fasy467-source.drv':
         specified: sha256-UmWI3NOE8Lf8ICHOR8nNpbCP9+g3R8XHRX+nUJsH6pY=
                     got:    sha256-mrUR84WY8BABb2+SqXJ/Q8ioIzvMgS9UZjzson5FPiU=

File: pkgs/development/tools/rust/itm/default.nix [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+  src = fetchFromGitHub {
+    owner = "rust-embedded";
+    repo = pname;
+    rev = "3e4251b42aa2e4b05ae372c47c7b835b8acae6dc";
+    sha256 = "15pa0ydm19vz8p3wairpx3vqzc55rp4lgki143ybgw44sgf8hraj";
+  };
+
+  cargoSha256 = "1d748vh1ig8r0cd70lsnw9fdmanic4jxxibbbagd5mrh8462v64v";
+  cargoPatches = [ ./itm_cargo_lock.patch ];
```
Comment (jonringer) [inline pkgs/development/tools/rust/itm/default.nix:19]: I'm curious why this is the necessary. Please upstream if it fixes the build.

File: pkgs/development/tools/rust/itm/default.nix [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+  meta = with lib; {
+    description = "A Rust crate and tool itmdump to parse and dump ARM ITM packets";
+    homepage = "https://github.com/rust-embedded/itm";
+    license = licenses.asl20;
+    maintainers = [ ];
```
Comment (jonringer) [inline pkgs/development/tools/rust/itm/default.nix:25]: Please add yourself as a maintainer when adding a new package.

https://nixos.org/manual/nixpkgs/stable/#reviewing-contributions-new-packages

File: pkgs/development/tools/rust/itm/default.nix (hunk 1 above)
Comment (davidarmstronglewis) [reply pkgs/development/tools/rust/itm/default.nix]: Without patching the Cargo.lock file I get build errors:

```
➜ nix-build -A itm
these derivations will be built:
  /nix/store/b95c43zmcs2kz1mwy145k7fzpqpy97yy-itm-0.3.1.drv
building '/nix/store/b95c43zmcs2kz1mwy145k7fzpqpy97yy-itm-0.3.1.drv'...
unpacking sources
unpacking source archive /nix/store/zz34rw0bgrwg840hmwd9c3mai01ivv5v-source
source root is source
Executing cargoSetupPostUnpackHook
unpacking source archive /nix/store/dm1rj9v3bz623726f3w10y93kg0521jx-itm-0.3.1-vendor.tar.gz
Finished cargoSetupPostUnpackHook
patching sources
Executing cargoSetupPostPatchHook
Validating consistency between /build/source//Cargo.lock and /build/itm-0.3.1-vendor.tar.gz/Cargo.lock
Finished cargoSetupPostPatchHook
configuring
building
Executing cargoBuildHook
++ env CC_x86_64-unknown-linux-gnu=/nix/store/s5hkav7whndbfz0szshpb46h4idqdq9a-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/s5hkav7whndbfz0szshpb46h4idqdq9a-gcc-wrapper-10.3.0/bin/c++ CC_x86_64-unknown-linux-gnu=/nix/store/s5hkav7whndbfz0szshpb46h4idqdq9a-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/s5hkav7whndbfz0szshpb46h4idqdq9a-gcc-wrapper-10.3.0/bin/c++ cargo build -j 4 --target x86_64-unknown-linux-gnu --frozen --release
error: the lock file /build/source/Cargo.lock needs to be updated but --frozen was passed to prevent this
If you want to try to generate the lock file without accessing the network, remove the --frozen flag and use --offline instead.
builder for '/nix/store/b95c43zmcs2kz1mwy145k7fzpqpy97yy-itm-0.3.1.drv' failed with exit code 101
error: build of '/nix/store/b95c43zmcs2kz1mwy145k7fzpqpy97yy-itm-0.3.1.drv' failed
```

I remember seeing prior art around patching the `Cargo.lock` file like this which is why I've gone this route. I think this happens when the `Cargo.lock` file is out of date when a release commit is tagged and figured I'd do the same here. I'll try and communicate the issue to the maintainers in the meantime.

File: pkgs/development/tools/rust/itm/default.nix (hunk 2 above)
Comment (davidarmstronglewis) [reply pkgs/development/tools/rust/itm/default.nix:25]: Will do!

File: maintainers/maintainer-list.nix [hunk 3]
```diff
@@ -2587,6 +2587,12 @@
     githubId = 91113;
     name = "David Kleuker";
   };
+  davidarmstronglewis = {
+    email = "davidlewis@mac.com";
+    github = "davidarmstronglewis";
+    githubId = 6754950;
+    name = "David Armstrong Lewis";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix:2595]: should be a separate commit

Comment (jonringer) [reply maintainers/maintainer-list.nix:2595]: Please make this into a separate commit:

```
maintainers: add <name>
```

File: pkgs/development/tools/rust/itm/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  src = fetchFromGitHub {
+    owner = "rust-embedded";
+    repo = "itm";
+    rev = "96d7710f2ba8f915668e95d67e0def961567c195";
+    sha256 = "15pa0ydm19vz8p3wairpx3vqzc55rp4lgki143ybgw44sgf8hraj";
+  };
+
+  cargoSha256 = "1d748vh1ig8r0cd70lsnw9fdmanic4jxxibbbagd5mrh8462v64v";
+  cargoPatches = [ ./cargo_lock.patch ];
```
Comment (jonringer) [inline pkgs/development/tools/rust/itm/default.nix:19]: What I was trying to say, is that it seems that upstream is out-of-date. The proper fix would be to upstream the lock update, then use `fetchpatch` to retrieve it from the PR. Instead of vendoring the patch into the nix expression.

File: pkgs/development/tools/rust/itm/default.nix (hunk 2 above)
Comment (davidarmstronglewis) [reply pkgs/development/tools/rust/itm/default.nix]: Added in #145441 which I've rebased onto.

File: maintainers/maintainer-list.nix (hunk 3 above)
Comment (davidarmstronglewis) [reply maintainers/maintainer-list.nix:2603]: Done in #145441

File: pkgs/development/tools/rust/itm/default.nix (hunk 4 above)
Comment (davidarmstronglewis) [reply pkgs/development/tools/rust/itm/default.nix:19]: So, after some digging I've learned that this tool only exists up to the 0.3.1 release of this tool. I want to package up this tool to make it easier for others following along with the Rust-Embedded Discovery using NixOS, but it looks like the CLI tool I'm trying to package has been extracted into another non-compatible tool [here](https://github.com/japaric/itm-tools).

The book still calls for this tool to be used (see https://docs.rust-embedded.org/discovery/03-setup/index.html#itmdump) so I think it's still appropriate to package this up at least until the Discovery book is reworked to reference this other tool.

Comment (davidarmstronglewis) [reply pkgs/development/tools/rust/itm/default.nix:19]: Uh, so, as for fixing this upstream, this tool isn't actually around anymore so there's nothing to fix upstream 😓

Comment (jonringer) [reply pkgs/development/tools/rust/itm/default.nix:19]: hmm, if it's no longer maintained, then I'm not really comfortable adding it to nixpkgs. because all ownership of maintenance will be put on us.

Comment (davidarmstronglewis) [reply pkgs/development/tools/rust/itm/default.nix:19]: Very fair. I will see what I can do :)

**Comment by davidarmstronglewis** [comment]
Closing this PR until the Rust Embedded Working Group finishes updating the Discovery book.

- https://github.com/rust-embedded/discovery/pull/414
- https://github.com/rust-embedded/itm/pull/41
